Eguage

huss opened the pull request:
# Description

This enables OED to use a pull system to acquire readings from eGauge meters. The work to incorporate eGauge meters was done by @lindavin where @huss only made changes to be compatible with other recent changes. It appears (for unknown reasons) that the commit trail indicating that @lindavin did most of this work was lost in a merge.

Unlike previous meters that we pull, eGauge uses a general URL and not just an IP address. As a result, references to IP were generally renamed URL in the code.

This code includes the recent changes in moment in PR #755 so it should be merged after that. The actual changes will be clearer after that.

## Type of change

- [x] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

None known

huss committed:
Davin and my work combined

huss committed:
ip - > url + more

- renames ip/ipaddress to url since OED now uses URL.
Note done with MAMAC and Obvius since internally they use ip.
- meter.displayable not shown if not admin since cannot see
those meters.

huss committed:
remove convertToReadings since not used

huss committed:
fix up validation message

- It tried to give an idaddress but that did not exist.
- Made message and sequence match other ones in pipeline

huss committed:
change from 0.8.0 to 1.0.0 for migration

The current plan is to have this be part of release 1.0.

huss committed:
package info consistent with remaining 0.8.0

huss commented:
The non-CodeQL testing failures are believed to be fixed with PR #755. Any remaining will be addressed after that PR clears and is integrated.

huss commented:
I dismissed the two CodeQL issues because the [eGauge documentation](https://docs.google.com/document/d/11l1SG2gHXuiQi_wpf6aroC4KjmTCIO_UtFdC--CJ5vQ/edit#heading=h.t94w0jkciyay) under Authentication, /auth Service, /auth/login in section Log in via Digest specifies that MD5 should be used in this process. Thus, OED is using what seems to be needed.

huss commented on src/server/models/eGauge/eGaugeRequestor.js:
```diff
@@ -0,0 +1,183 @@
+			const readingValues = dataRange.rows;
+			// TODO what is the 's' parameter for? I removed it and it still works.
+			const endTs = moment.unix(dataRange.ts).tz(timezoneUse);
+
```
Should timeDelta be the same as timeStep (if all went well)?

lindavin replied:
Yes, they should be the same. It might be useful to add a sanity check to ensure that eGauge is sending data at the correct time steps, but I don't think it would be necessary since eGauge seems to be working well.

huss commented on src/server/models/eGauge/eGaugeRequestor.js:
```diff
@@ -0,0 +1,183 @@
+		}
+		const meterReadings = [];
+		const readingType = data.registers[0].type;
+		
```
Was it okay to remove the s parameter?

lindavin replied:
Yes, it's okay to remove the parameter. The 's' parameter was just to tell moment.unix that the input is the number of seconds after the Unix Epoch. I included it for clarity, but since this is the default parameter value it's okay to remove it.

huss committed:
modify comments

huss reviewed (commented):
My review found this looks good. A couple of questions in my comments for @lindavin to look at. I pushed some updates that mostly modify the comments in the code. I also decided this will stay v0.8.0 and be part of v1.0.0 update. Final testing will be done after the moment PR is merged and the final merge is done to this PR.

lindavin commented on package.json:
```diff
@@ -83,6 +84,7 @@
     "es6-promisify": "~5.0.0",
     "express": "~4.16.4",
     "express-rate-limit": "~5.5.1",
+    "fast-xml-parser": "~4.0.1",
```
I believe fast-xml-parser is not being used in eGauge; although it was at first. So it can removed from package.json and package-lock.json.

huss replied:
I could not find any usage either so I removed.

lindavin commented on src/server/app.js:
```diff
@@ -57,6 +57,7 @@ if (log.level !== LogLevel.SILENT) {
 
 app.use(favicon(path.join(__dirname, '..', 'client', 'public', 'favicon.ico')));
 app.use(bodyParser.json({ limit: '50mb' }));
+app.use(bodyParser.raw({ limit: '50mb', type: ['application/xml'] }));
```
This can also be removed, because we are not using xml.

huss replied:
Done.

lindavin commented on src/server/models/Meter.js:
```diff
@@ -177,15 +179,15 @@ class Meter {
 
 		this.name = name;
-		this.ipAddress = ipAddress;
+		this.url = rul;
```
Is "rul" a typo?

huss replied:
Fixed. I note the same fix is in PR #781 where it was also found.

lindavin reviewed (approved):
Thank you @huss for the hard work! I left a couple comments, that I think should be addressed. Besides the merge conflict, overall, this PR looks good to me.

huss committed:
address PR comments

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/huss/764

huss commented:
Thanks to @lindavin for the review. I have addressed all the comments. I also merged in development to resolve the merge conflict. Testing indicates it is getting all the meter data correctly (expect for fix for daylight savings in PR #781). This is ready to merge.
